src/clj_puppetdb/http.clj

```diff
@@ -24,47 +24,86 @@
   `puppetlabs.ssl-utils.core/pems->ssl-context` function."
   [:ssl-cert :ssl-key :ssl-ca-cert])
 
+(defn- make-client-common
+  [^String host opts]
+  (let [opts (assoc opts :as :stream)
+        info {:host host}
+        info (if-let [ssl-context (:ssl-context opts)]
+               (assoc info :ssl-context ssl-context)
+               info)]
+    (fn
+      ; This arity-overloaded variant of the function is the one called by the `GET` function.
+      ; The `this` parameter refers under normal circumstances to this very function and it is
+      ; needed to be able to call another arity-overloaded variant of the function from the
+      ; function body. When the `this` parameter refers to a different function then it
+      ; effectively overrides the other arity-overloaded variants of the function.
+      ([this path params]
+       (let [query (-> (StringBuilder. host)
+                       (.append path))]
+         (if-not (empty? params)
+           (-> query
+               (.append \?)
+               (.append (-> params
+                            q/params->json
+                            map->query))))
+         ; now call the single argument variant of the function to execute the query
+         (this (.toString query))))
```

Comment:
Is there any reason to think clojure.string/join is less performant?
Arguably we shouldn't optimize for speed above all else anyway. I think
there's value in idiomatic code and simplicity even at the cost of some
degree of outright speed.

On 2 April 2015 at 14:15, mruzicka <notifications@github.com> wrote:

> In src/clj_puppetdb/http.clj
> <https://github.com/puppetlabs/clj-puppetdb/pull/11#discussion_r27655014>:
>
> > +      ; This arity-overloaded variant of the function is the one called by the `GET` function.
> > +      ; The `this` parameter refers under normal circumstances to this very function and it is
> > +      ; needed to be able to call another arity-overloaded variant of the function from the
> > +      ; function body. When the `this` parameter refers to a different function then it
> > +      ; effectively overrides the other arity-overloaded variants of the function.
> > +      ([this path params]
> > +       (let [query (-> (StringBuilder. host)
> > +                       (.append path))]
> > +         (if-not (empty? params)
> > +           (-> query
> > +               (.append \?)
> > +               (.append (-> params
> > +                            q/params->json
> > +                            map->query))))
> > +         ; now call the single argument variant of the function to execute the query
> > +         (this (.toString query))))
>
> The intent here is optimization. While admittedly not very clojury, it is
> as straightforward as reasonably possible. Moreover the (str) function
> uses internally StringBuilder too.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/puppetlabs/clj-puppetdb/pull/11/files#r27655014>.
>